api/src/main/java/org/apache/iceberg/BaseInputSplit.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+public class BaseInputSplit<T extends ScanTask> implements InputSplit<T> {
+  private final List<T> tasks;
+
+  public BaseInputSplit(List<T> tasks) {
+    Preconditions.checkNotNull(tasks, "tasks cannot be null");
+    this.tasks = Collections.unmodifiableList(Lists.newArrayList(tasks));
```

Comment:
I need to double check Kryo serialization. Maybe, I should return an unmodifiable wrapper only in `files()`.

Reply:
Yeah, for serialization we usually want to go with a Java array.

Reply:
Doesn't Kryo support mutable collections? I remember having issues with immutable collections only.

I was reluctant to use an array because of generics. I will most likely will need to pass `Class<T>` around.

Reply:
Added tests to cover mutable lists but I generally don't mind passing `Class<T>` to switch to arrays.